src/XMakeBuildEngine/Evaluation/LazyItemEvaluator.cs

```diff
@@ -0,0 +1,704 @@
+
+                            foreach (ProjectMetadataElement metadatumElement in _metadata)
+                            {
+#if FEATURE_MSBUILD_DEBUGGER
+                                //if (DebuggerManager.DebuggingEnabled)
+                                //{
+                                //    DebuggerManager.PulseState(metadatumElement.Location, _itemPassLocals);
+                                //}
+#endif
+
+                                if (!EvaluateCondition(metadatumElement, ExpanderOptions.ExpandAll, ParserOptions.AllowAll, _expander, _lazyEvaluator))
+                                {
+                                    continue;
+                                }
+
+                                string evaluatedValue = _expander.ExpandIntoStringLeaveEscaped(metadatumElement.Value, ExpanderOptions.ExpandAll, metadatumElement.Location);
+
+                                item.SetMetadata(metadatumElement, evaluatedValue);
+                            }
+                        }
+
+                        // End of legal area for metadata expressions.
+                        _expander.Metadata = null;
+                    }
+
+                    // End of pseudo batching
+                    ////////////////////////////////////////////////////
+                    // Start of old code
+                    else
+                    {
+                        // Metadata expressions are allowed here.
+                        // Temporarily gather and expand these in a table so they can reference other metadata elements above.
+                        EvaluatorMetadataTable metadataTable = new EvaluatorMetadataTable(_itemType);
+                        _expander.Metadata = metadataTable;
+
+                        // Also keep a list of everything so we can get the predecessor objects correct.
+                        List<Pair<ProjectMetadataElement, string>> metadataList = new List<Pair<ProjectMetadataElement, string>>();
+
+                        foreach (ProjectMetadataElement metadatumElement in _metadata)
+                        {
+                            // Because of the checking above, it should be safe to expand metadata in conditions; the condition
+                            // will be true for either all the items or none
+                            if (!EvaluateCondition(metadatumElement, ExpanderOptions.ExpandAll, ParserOptions.AllowAll, _expander, _lazyEvaluator))
+                            {
+                                continue;
+                            }
+
+#if FEATURE_MSBUILD_DEBUGGER
+                        //if (DebuggerManager.DebuggingEnabled)
+                        //{
+                        //    DebuggerManager.PulseState(metadatumElement.Location, _itemPassLocals);
+                        //}
+#endif
+
+                            string evaluatedValue = _expander.ExpandIntoStringLeaveEscaped(metadatumElement.Value, ExpanderOptions.ExpandAll, metadatumElement.Location);
```

Comment:
Can it happen for metadata strings to reference items that are not in ReferencedItems?

```
<Foo include="1;2;3">
<Bar Include="a">
   <m>@(Foo)<m>
<Bar>
```